### PR TITLE
Feature clustering model

### LINKED ISSUE #42: Algorithms of clustering and classification.
**Is your feature request related to a problem? Please describe.**
Data is not suitable for standard solutions by prediction algorithms and clustering / classification may help

**Describe the solution you'd like**
Explore algorithms of clustering and classification. Try to analyze raw data. reformat it, to use categorical prediction algorithms.

**Describe alternatives you've considered**
Maybe prediction algorithms will not be needed and to determine the goods that will be bought, there will be enough analysis.

**Additional context**
Look for similar tasks and their solutions. It is possible to use libraries.

### PR DESCRIPTION
**Pull request by ptrdiff**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

*Any change needs to be discussed before proceeding. Failure to do so may result in the rejection of the pull request.*

Please provide enough information so that others can review your pull request:

- Add clustering model with KMeans algorithm.
- Add configuration settings for this algorithm.

Explain the **details** for making this change. What existing problem does the pull request solve?

It was suggested to try clustering to identify the customer's kludge characteristics. The kmeans algorithm was used with a breakdown into 3 clusters. Which showed accuracy slightly more linear regression, but at the same time higher speed of operation.

**Closing issues**
Closes #42
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/robot-lab/tinkoff-optimization-of-procurement/pull/52%23discussion_r212837233%22%2C%20%22https%3A//github.com/robot-lab/tinkoff-optimization-of-procurement/pull/52%23discussion_r212921048%22%5D%2C%20%22comments%22%3A%20%7B%22Pull%2066ac6824ba636df8c7aaf61347176a34d2084776%20mlalgorithms/models/clustering_model.py%2027%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/robot-lab/tinkoff-optimization-of-procurement/pull/52%23discussion_r212837233%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Please%2C%20fix%20identation.%22%2C%20%22created_at%22%3A%20%222018-08-26T19%3A54%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/21951872%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Vasar007%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%22%2C%20%22created_at%22%3A%20%222018-08-27T09%3A40%3A46Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/21951872%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Vasar007%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20mlalgorithms/models/clustering_model.py%3AL1-44%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [x] <a href='#crh-comment-Pull 66ac6824ba636df8c7aaf61347176a34d2084776 mlalgorithms/models/clustering_model.py 27'></a> &nbsp;<b><a href='https://github.com/robot-lab/tinkoff-optimization-of-procurement/pull/52#discussion_r212837233'>File: mlalgorithms/models/clustering_model.py:L1-44</a></b>
- <a href='https://github.com/Vasar007'></a> Please, fix identation.

<a href='https://www.codereviewhub.com/robot-lab/tinkoff-optimization-of-procurement/pull/52?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/robot-lab/tinkoff-optimization-of-procurement/pull/52?approve=1'></a>&nbsp;<a href='https://github.com/robot-lab/tinkoff-optimization-of-procurement/pull/52'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by ptrdiff** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/robot-lab/tinkoff-optimization-of-procurement/pull/52'>Refresh</a>.<a href='#crh-update-none'></a>

File: mlalgorithms/models/clustering_model.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+                self.orders[persons_id] += np.array(label)
+
+        self.clustering_table = pd.DataFrame(
+            self.kmeans.fit_predict(
+                pd.DataFrame.from_dict(self.orders, orient='index')),
+            columns=['cluster'])
+
+        cluster_id = self.clustering_table['cluster'].value_counts().index[0]
+        larg_clust_center = self.kmeans.cluster_centers_[cluster_id]
+        self.largest_cluster_goods = (larg_clust_center >= 6).astype(np.int)
```
Comment (Vasar007) [inline mlalgorithms/models/clustering_model.py]: What is this magic number `6`? You can create constant for this at least. With our architecture now we cannot separate model class parameters from sklearn class parameters. Maybe we need to add this option?

File: mlalgorithms/models/clustering_model.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        for persons_id, label in zip(persons_ids, train_labels):
+            if self.orders.get(persons_id) is None:
+                self.orders[persons_id] = np.array(label)
+            else:
+                self.orders[persons_id] += np.array(label)
+
+        self.clustering_table = pd.DataFrame(
+            self.kmeans.fit_predict(
+                pd.DataFrame.from_dict(self.orders, orient='index')),
+            columns=['cluster'])
```
Comment (Vasar007) [inline mlalgorithms/models/clustering_model.py]: Please, fix identation.

File: mlalgorithms/models/clustering_model.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        larg_clust_center = self.kmeans.cluster_centers_[cluster_id]
+        self.largest_cluster_goods = (larg_clust_center >= 6).astype(np.int)
+
+    def predict(self, samples, **kwargs):
+        predictions = []
+        for i in samples:
+            if i[0] in self.clustering_table.index:
+                cluster_id = self.clustering_table.at[i[0], 'cluster']
+                clust_center = self.kmeans.cluster_centers_[cluster_id]
+                prediction = (clust_center >= 6).astype(np.int)
```
Comment (Vasar007) [inline mlalgorithms/models/clustering_model.py]: Same issue with magic number as I have written just now.

File: mlalgorithms/models/clustering_model.py (hunk 2 above)
Comment (Vasar007) [inline mlalgorithms/models/clustering_model.py]: :+1:

**Pull request merged**